### ISSUE TITLE
Change ->dir interface.

### ISSUE THREAD

**Issue by toddr** [cpanel/Test-MockFile#69]
To be more consistent, we want to change opendir/glob to simply go off of the list of mocked files.

```perl
my $d = Test::MockFile->dir('/etc', ['foo', 'bar']); # DIES this is now illegal.

my $dir = '/path';
my $f = $m->dir($dir ) # Mocked but not there.
-d '/dirpath'; # False
mkdir($dir, 0755)
-d '/dirpath'; # True
```
dir now no longer takes a second arg or it dies.

opendir/read and glob should now go off of the mocked files or mocked + existing when not in strict mode.

**Comment by xsawyerx** [comment]
(I don't understand the comment on "mocked + existing.")

The problem in the interface is that `->file()` and `->dir()` methods do two different things:
* Take over a path to be able to return results for it (including nonexistence)
* Provide content within those paths (`->file()` for file content and `->dir()` for "directory contents" - notice the quotes dripping with sarcasm and annoyance)

With `->file()` this is fairly simple. If you give nothing or `undef`, then that's nonexistence. Anything else is existence + content.

But with `->dir()` there is a problem because there are three modes for a directory: 
1. Nonexistent directory (no content)
2. Existent but empty directory (still contains `.` and `..`)
3. Existent and containing content that was created through `Test::MockFile->file()` (assuming we removed creating fake and buggy content with `->dir( $name, [...] )`)

The question is how to allow the user to control between these three. `->dir( $name, $stats )` seems to make "take over this" like with `->file($name)` but what happens if the user wants to say "I want this directory to exist but be empty"? They can create an empty fake file `->file("$dir/.exists")` as people do with Git (`->file("$dir/.gitignore")`) but I think that's counterintuitive and the logical code will have to skip those files.

I suggest the following:
* Take control of paths with `->file()` and `->dir()`
* Set file existence and content with either `open` or `->file( $name => $content || undef )`
* Calling `->dir()` on a directory of files (created with `->file()` or `open()`) will reflect those files (along with `.` and `..`)
* Calling `->dir()` when no files exist (none created with `->file()` or `open()`) will take over a directory and show it as nonexistent
* Create empty directories using `mkdir`

```
# nonexistent dir
$d = Test::MockFile->dir('/etc')->contents(); # undef

# retrieve directory content after creating content with ->file()
$f = Test::MockFile->file('etc/foo');
$d = Test::MockFile->dir('/etc')->contents(); # . .. foo

# retrieve directory content for empty dir 
# this already works
my $d = Test::MockFile->dir('/etc');
mkdir '/etc';
$d->contents(); # . ..
```

**Comment by xsawyerx** [comment]
I'm now also realizing a can of worms with `stat`.

Here is what I would expect:

```
my @stat_as_array;
my %stat = stat_as_hash(@stat_as_array);
my $d = Test::MockFile->dir( '/etc/' => \%stat );
$d->contents(); # undef
mkdir $d;
$d->contents(); # . ..
my @new_stat = stat '/etc/';
is_deeply( \@new_stat, \@stat_as_array ); # 1
```

In other words, just because you took over a directory with `stat` information, it doesn't mean it exists. Once you create it, it will exist with `stat` information (the default or the one you gave). However, here lie dragons. Think of keywords that change that `stat` info, like `mkdir()` or `readdir()` which change the `atime` or `ctime`. It makes sense to say "stats are always what you provided," but... we still want to support other keywords that change the stats (like `chown()` in #15) so we're not really consistent here.

Example of this being a bit confusing:

```
# ctime stays what we gave as the stat value
my $ctime = time; # 1641483380
$d = Test::MockFile->dir( '/etc', { 'ctime' => $ctime, 'gid' => $> + 1 } );
sleep 5;
mkdir $d;
say ( stat '/etc' )[10]; # $ctime

# with chown from #15 implemented
say ( stat $f )[5]; # $> + 1 == 1001
chmod $<, $>, '/etc/foo'; # change gid from 1001 to 1000
say ( stat $f )[5]; # 1000

# and yet, readdir does not change atime
my $atime_before = ( stat '/etc' )[8];
opendir my $dh, '/etc' or die;
readdir $dh;
closedir $dh;
say ( stat '/etc' )[8] # $atime_before
```

So which keywords are allowed to change the `stat` and which aren't and why?

If we decide that `mkdir $dirname` *does* set the `ctime`, then does it override the `ctime` provided on the `->dir()` method call? If it doesn't, does it override the default `ctime`? And if so, then why even have a default `ctime` if you both need to call `mkdir` to retrieve a `ctime` and by calling `mkdir()` you override the default `ctime`?

Some options:
1. There's a limited list of keywords that change the stats (like `chown()`) that are documented. The rest do not. (What if someone *does* want to change a particular `stat` entry if the keywords that change it won't actually change it? "Bug report: I've tried `opendir()` but `atime` didn't change!")
2. All keywords change whatever they should. Here be even bigger dragons. There's *a lot* to account for with these without even getting into mounting options (`mount -o noatime -o nodiratime`).
3. Do not support keywords that only change the `stat` because `stat` can only be created during `->file()` and `->dir()` calls (not good for #15 and any ticket that wants to check permissions).
4. Draw a line at some capabilities like permissions and - other than those specific ones, keywords do not change `stat`. (Will we provide an interface for changing that information or just tell people to create new objects? New objects might be easier.)

**Comment by toddr** [comment]
The interface:
```perl
my $m = 'Test::MockFile';
$m->file('/foo', 'djhdjhdf', {perms => 0});


```

**Comment by xsawyerx** [comment]
Resolved when #69 was merged.
